simple postprocess cookbook

jaustermann opened the pull request:
Here is a very simple cookbook to introduce dynamic topography and the geoid. @Shangxin-Liu @ian-r-rose let me know whether you'd like me to adjust the text. 
I will update the 'real Earth' cookbook with figures of the bottom topography and geoid once my runs have completed. These will be a bit more exciting than just a harmonic perturbation :)

ian-r-rose commented on doc/manual/manual.bib:
```diff
@@ -12,7 +12,19 @@ @string{ajs
 @string{pnas = {Proceedings of the National Academy of Sciences}}
 
-
+@Article{ZGH93,
```
Thanks for adding this.

ian-r-rose commented on doc/manual/manual.tex:
```diff
@@ -7101,6 +7101,66 @@ \subsubsection{Simple convection in a spherical 3d shell}
+topography and the geoid. We initialize the model with a harmonic 
+perturbation of degree 4 and order 2 and calculate the
+instantaneous solution. Analogous to the previous setup we use a spherical
+shell Geometry model and a simple Material model. 
```
Do we have a convention on referring to material models and geometry models? The capitalization clanged in my ears.

ian-r-rose commented on doc/manual/manual.tex:
```diff
@@ -7101,6 +7101,66 @@ \subsubsection{Simple convection in a spherical 3d shell}
+\lstinputlisting[language=prmfile]{cookbooks/shell_3d_postprocess/shell_3d_postprocess.part.prm.out}
+
+This initial condition results in distinct flow cells that cause local up- and
+downwellings (Figure~\ref{fig:pp}). This flow perturbs the top and bottom
```
A bit more detail here might be nice. Something like
> This flow deflects the top and bottom boundaries of the mantle away from their reference height,
> a process known as dynamic topography. The deflection of the surface also causes a perturbation
> in the gravitational field of the planet.

ian-r-rose commented on doc/manual/manual.tex:
```diff
@@ -7101,6 +7101,66 @@ \subsubsection{Simple convection in a spherical 3d shell}
+\paragraph{Dynamic topography at the surface and core mantle boundary.}
+Dynamic topography is calculated at the surface and bottom of the domain 
+through a stress balancing approach where we assume that the radial stress at 
+the surface is balanced by excess topography. We use a consistent boundary flux 
```
"excess (or deficit) topography" ?

ian-r-rose replied:
"...the consistent boundary flux (CBF) method"

ian-r-rose commented on doc/manual/manual.tex:
```diff
@@ -7101,6 +7101,66 @@ \subsubsection{Simple convection in a spherical 3d shell}
+through a stress balancing approach where we assume that the radial stress at 
+the surface is balanced by excess topography. We use a consistent boundary flux 
+method to calculate the radial stress at the surface \cite{ZGH93}. For the 
+bottom surface we chose a convention that positive values are up (out)
```
perhaps "we define positive values as up" instead of "we choose a convention that positive values are up" ?

ian-r-rose commented on doc/manual/manual.tex:
```diff
@@ -7101,6 +7101,66 @@ \subsubsection{Simple convection in a spherical 3d shell}
+should always be zero, however, the outputted values might not fullfill this 
+constraint if the resolution of the model is not high enough to provide an
+accurate solution. This cookbook only uses a refinement of 2, which is relatively
+low resolution. 
```
I think this is good.

ian-r-rose commented on doc/manual/manual.tex:
```diff
@@ -7101,6 +7101,66 @@ \subsubsection{Simple convection in a spherical 3d shell}
+
+\paragraph{Geoid anomalies.}
+Geoid anomalies are perturbations of the gravitational equipotential surface that 
+are due to density variations within the mantle as well as  deflections of the 
```
Two spaces (though I suppose that does not matter in LaTeX).

ian-r-rose commented on doc/manual/manual.tex:
```diff
@@ -7101,6 +7101,66 @@ \subsubsection{Simple convection in a spherical 3d shell}
+\begin{figure}
+  \includegraphics[width=\textwidth]{cookbooks/shell_3d_postprocess/postprocess_cookbook-01.png}
+  \hfill
+  \caption{\it Panel a shows an equatorial cross section of the temperature distribution and 
```
Panel (a), Panel (b), (c), (d) for clarity

jaustermann commented:
I think I addressed @ian-r-rose's comments. I also updated the 'real Earth' cookbook to include output from the CBF dynamic topography postprocessor and the geoid postprocessor. I realize now it would have probably been easier to review this if I had made this update a separate pull request. I'm sorry. Anyhow, I think this is ready to be looked at again.

gassmoeller commented on cookbooks/shell_3d_postprocess.prm:
```diff
@@ -0,0 +1,107 @@
+# We assume fixed temperatures and free slip at the inner and outer boundary
+subsection Model settings
+  set Fixed temperature boundary indicators   = 0,1
+  set Tangential velocity boundary indicators = 0,1
```
Could you replace the boundary numbers by the names? We try to phase out the use of the numbers. Also you can probably already use the new names "bottom,top" instead of "inner,outer" (if not you might need to rebase to a current master). The old names still work, but we try to use the new ones, because they are geometry independent.

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -7101,6 +7101,73 @@ \subsubsection{Simple convection in a spherical 3d shell}
+\textit{This section was contributed by Jacqueline Austermann, Ian Rose and Shangxin Liu}
+
+
+There are several postprocessors that can be used to turn the the velocity 
```
double the

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -7101,6 +7101,73 @@ \subsubsection{Simple convection in a spherical 3d shell}
+consistent boundary flux (CBF) method to calculate the radial stress 
+at the surface \cite{ZGH93}. For the 
+bottom surface we define positive values as up (out)
+and negative values are in (down), analogous to the deformation of the upper 
```
I think you mixed up the terms in and outside of the parentheses.

bangerth replied:
@gassmoeller : Hm, I think she means "out of the core" and "into the core", which corresponds to "up" and "down".

Or did you mean to say that symmetry suggests that if you use "up (out)" in the first line, you should use "down (in)" in the second?

jaustermann replied:
I corrected it to "positive values as up (out) and negative values are down (in)"

gassmoeller replied:
Yes, that is what I meant.

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -7101,6 +7101,73 @@ \subsubsection{Simple convection in a spherical 3d shell}
+equator causes alternating topography high and lows at the top and
+bottom surface (Figure~\ref{fig:pp}). 
+In Figure~\ref{fig:pp} c, d we have subtracted the mean dynamic topography from
+the output field. Since mass is conserved within the Earth, the mean dynamic topography
```
I thought the option to remove the mean is gone? Is this still the case or did you write it for an earlier version of the postprocessor?

jaustermann replied:
Yes, the option is gone. I calculated the mean from the output and then subtracted it outside of aspect. I can either make that clear in the text or simply don't subtract the mean. I'm leaning towards the latter (that makes it consistent with the next cookbook) but let me know what you think.

gassmoeller replied:
I do not feel strongly either way. Choose the way you prefer.

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -7101,6 +7101,73 @@ \subsubsection{Simple convection in a spherical 3d shell}
+are due to density variations within the mantle as well as deflections of the 
+surface and core mantle boundary. The geoid postprocessor subtracts the mean
+dynamic topography at the top and bottom surface when evaluating its contribution
+to the gravity field. 
```
Same question as above. Do we still subtract the mean?

jaustermann replied:
Good catch, yes and no. I previously looked at the geoid postprocessor and it says in line 271 "Subtract the average dynamic topography" so I assumed that's what is done. Looking more carefully at it now it seems that this isn't actually done (probably a remnant from an earlier version). But, the default minimum degree is 2 so that removes the mean. I think that's what's going on, does that sound right? If so I can clarify this in the text.

bangerth replied:
@ian-r-rose ?

gassmoeller replied:
That sounds about right. I think the comment is a relic from the first version that used its own dynamic topography calculation.

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -7195,43 +7262,83 @@ \subsubsection{3D convection with an Earth-like initial condition}
 with upper mantle features such as high heat flow at mid ocean ridges and low
 heat flow at cratons is correctly initialized by the tomography model. 
+The mantle flow and bouyancy variations produce dynamic topography on the 
+top and bottomw surface, which is shown for 2 refinement steps (\ref{fig:ic-2}a and c, respectively) 
```
bottom

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -7195,43 +7262,83 @@ \subsubsection{3D convection with an Earth-like initial condition}
+(for a given refinement, the resolution at the core mantle boundary is
+higher than the resolution at the surface). The geoid postprocessor removes the
+average top and bottom surface when evaluating its contribution
+to the gravity field. The average geoid height is therefore zero. 
```
Is this still the case now that it uses the new dynamic topography postprocessor?

gassmoeller commented:
/run-tests

bangerth commented on doc/manual/manual.tex:
```diff
@@ -7101,6 +7101,73 @@ \subsubsection{Simple convection in a spherical 3d shell}
 
+\subsubsection{Postprocessing spherical 3D convection}
+\label{sec:cookbooks-shell_3d_postprocess.prm}
+\textit{This section was contributed by Jacqueline Austermann, Ian Rose and Shangxin Liu}
```
I know you're in Cambridge, but maybe we can convince you to use the Oxford comma anyway? ;-)

bangerth commented on doc/manual/manual.tex:
```diff
@@ -7101,6 +7101,73 @@ \subsubsection{Simple convection in a spherical 3d shell}
+topography and the geoid. We initialize the model with a harmonic 
+perturbation of degree 4 and order 2 and calculate the
+instantaneous solution. Analogous to the previous setup we use a spherical
+shell geometry model and a simple material model. 
```
Can you add a sentence somewhere that provides the path to the complete .prm file, as is done in most other cookbooks?

bangerth commented on doc/manual/manual.tex:
```diff
@@ -7101,6 +7101,73 @@ \subsubsection{Simple convection in a spherical 3d shell}
+at the surface \cite{ZGH93}. For the 
+bottom surface we define positive values as up (out)
+and negative values are in (down), analogous to the deformation of the upper 
+surface. Dynamic topography can be outputted in text format (which writes the 
```
I've heard the term "outputted" several times at the hackathon, even from native speakers, but it always strikes me as wrong because the past tense of "put" is "put", and so the past tense of "output" should be "output". No? I think we need a native speaking tiebreaker -- @ian-r-rose , @ljhwang ?

jaustermann replied:
I've asked some native speakers around my office and consulted the internet; sounds like both are correct ("I've output xyz" and "I've outputted xyz") so if there are no objections I'll leave it as is :)

bangerth replied:
Yes, sure :-)

bangerth commented on doc/manual/manual.tex:
```diff
@@ -7101,6 +7101,73 @@ \subsubsection{Simple convection in a spherical 3d shell}
+In this model, downwellings coincide with lows in dynamic topography. That 
+means the mass deficit caused by the depression at the surface is not 
+fully compensated by the high density material below the depression that 
+drags the surface down. The geoid postprocessor uses a spherical harmonic 
```
I don't understand this. Wouldn't both downflow and dense material combine for a low in dynamic topography? Or do you mean "low in the geoid equisurface" in line 7155?

jaustermann replied:
Yes I meant low in the geoid anomaly. Thanks for catching that!

bangerth commented:
In the place where you first mention the 4 refinement figures, can you add a footnote that says something "Computations with 4 global refinements are expensive, and consequently this is not the default for this cookbook. For example, as of 2017, it takes XXX cores YYY hours of walltime to finish this cookbook with 4 global refinements." (assuming you still know the numbers).

Also -- great cookbook! Thanks for putting it together!

jaustermann commented:
Thanks for your comments Wolfgang, they are all in my branch now (including the Oxford comma ;). I'll wait until I have all comments resolved before I push again. The failed test should go away once I rebase to master. Regarding the computation time, on the Harvard cluster it took 2h on 64 cores, where 90% of the time is spent on setting up the initial condition (ouch!).

bangerth commented:
2h @ 64 core -- that's not too bad. I had feared worse. But I think it would still be useful to tell readers why this isn't the default.

jaustermann commented:
Yep, I added what you wrote and filled in the numbers.

jaustermann commented:
Sorry it took me so long to get back to this, but I think I addressed all the comments so this should be ready for another look.

gassmoeller commented on cookbooks/shell_3d_postprocess.prm:
```diff
@@ -0,0 +1,106 @@
+# the instantaneous solution.
+set End time                               = 0
+set Use years in output instead of seconds = true
+set Output directory                       = output_s2
```
usually cookbooks use an output directory of the form: output-prmname, i.e. output-shell_3d_postprocess in this case. This is mostly to avoid cookbooks overwriting each other, which should not happen with your name. Still for conformity please update.

gassmoeller commented on doc/modules/changes/20170612_austermann:
```diff
@@ -0,0 +1,5 @@
+New: Added a cookbook that uses the dynamic topography and geoid postprocessor in a simply
```
`simple` instead of `simply`

jaustermann commented:
Fixed the typos and squashed it with the previous commit. Thanks @gassmoeller, @bangerth, and @ian-r-rose for looking through this!

gassmoeller commented:
Thanks!
